resources/assets/coffee/react/beatmaps.coffee

```diff
@@ -0,0 +1,259 @@
+      method: 'get'
+      dataType: 'json'
+      data: $.extend(@buildSearchQuery(searchText), 'page': @state.paging.page + 1)
+    .done (data) =>
+      more = data['data'].length > 10
+      newState =
+        beatmaps: @state.beatmaps.concat(data['data'])
+        paging:
+          page: @state.paging.page + (if more then 1 else 0)
+          url: @state.paging.url
+          more: more
+        loading: false
+
+      @setState newState
+
+  showLoader: =>
+    @setState loading: true
+    $('#loading-area').show()
+
+  hideLoader: =>
+    @setState loading: false
+    $('#loading-area').hide()
+
+  updateFilters: (_e, payload) =>
+    newFilters = $.extend({}, @state.filters) # clone object
+    newFilters[payload.name] = payload.value
+
+    if @state.filters != newFilters
+      @setState filters: newFilters, ->
+        $(document).trigger 'beatmap:search:start'
+
+  updateSort: (_b, payload) =>
+    if @state.sorting != payload
+      @setState sorting: payload, ->
+        $(document).trigger 'beatmap:search:start'
+
+  convertFilterKeys: (obj) ->
+    newObj = {}
+    for key of obj
+      newObj[@charToKey(key)] = obj[key]
+    return newObj
+
+  restoreState: ->
+    state = {}
+    params = location.search.substr(1).split('&')
+    filters = []
+    sorting = {'field': 'ranked', 'order': 'desc'}
+
+    for filter of @getFilterState()
+      filters.push(filter)
+
+    expand = false
+    present = false
+    query = null
+
+    for part in params
+      [key, value] = part.split('=')
+      if $.inArray(key, ['g', 'l', 'e', 'r']) > -1
+        expand = true
+
+      if $.inArray(key, filters) > -1
+        state[key] = value
+        present = true
+
+      if key == 'so'
+        [sort_field, sort_order] = value.split('_')
+        if $.inArray(sort_field, @validSortFields()) > -1 && $.inArray(sort_order, @validSortOrders()) > -1
+          sorting = {'field': sort_field, 'order': sort_order}
+          present = true
+
+      if key == 'q'
+        $('#searchbox').val(value)
+        query = value
+        present = true
+
+    if present
+      newState = $.extend(this.getFilterState(), state) # clone
+      newState = @convertFilterKeys(newState)           # convert keys
+
+      @setState filters: newState, sorting: sorting, just_restored: true, ->
+        @search()
+        if expand
+          $('#search').addClass 'expanded' #such seperation of concerns
+
+  componentDidMount: ->
+    @componentWillUnmount()
+    $(document).on 'beatmap:load_more', @loadMore
+    $(document).on 'beatmap:search:start', @search
+    $(document).on 'beatmap:search:done', @hideLoader
+    $(document).on 'beatmap:search:filtered', @updateFilters
+    $(document).on 'beatmap:search:sorted', @updateSort
+    $(document).on 'ready page:load osu:page:change', (e) =>
+      @restoreState()
+    $(window).on 'popstate', (e) =>
+      @restoreState()
+
+  componentWillUnmount: ->
+    $(document).off 'beatmap:load_more'
```

Comment:
namespace everything by suffixing event name with `.beatmapListing` and enjoy cheap debinding with single `$(document).off '.beatmapListing'` woo :dancer: